feat: Show next available day/hour and minutes on widget

iamsivin opened the pull request:
# Pull Request Template

## Description

Fixes https://linear.app/chatwoot/issue/CW-1283/show-next-available-dayhour-on-widget

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
https://www.loom.com/share/b3dfc9c2dc9445c1a5262bc64db07a0c


## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

pranavrajs commented:
@iamsivin Nice work! A couple of comments.

1/ Let us only show the minutes if the time is less than an hour. (For, eg, Instead of 4 hr 10 minutes, let us show it as 5hrs.) If the time is less than 1 hr, we can show that we will be back online in X minutes.

2/ Instead of 4 days, we can show that we will be back online on Tuesday. If it is tomorrow, we can just show that we will be back online tomorrow.

These would be easier to read and understand. Let me know your thoughts.

iamsivin commented:
> @iamsivin Nice work! A couple of comments.
> 
> 1/ Let us only show the minutes if the time is less than an hour. (For, eg, Instead of 4 hr 10 minutes, let us show it as 5hrs.) If the time is less than 1 hr, we can show that we will be back online in X minutes.
> 
> 2/ Instead of 4 days, we can show that we will be back online on Tuesday. If it is tomorrow, we can just show that we will be back online tomorrow.
> 
> These would be easier to read and understand. Let me know your thoughts.

@pranavrajs Done. Can you take a look?

1.  Now (eg: 2hr 10min) then, it will show as "We will be back online in 3 hours"

2. Now if there is a 2-day difference we show the exact day name like "We will be back online on Monday".  if there is only a 1-day difference, then we show "We will be back online tomorrow"

scmmishra reviewed (commented):
One UX suggestion if the number of hours is greater than 3, we can just show the time at which the team will be back online "We will be back at 9:00 AM" 

Doing minute level calculations in the head will be a bit difficult, say right now it's `7:38 AM` and we say, "we will be back in 2 hours 23 minutes" Doing this calculation in the head is a lot of friction. 

Also we could keep neater intervals to show good numbers instead of showing each minute.

Say if the agents will be online in `3 hours 48 minutes`. We can round it to the next 10 or 30 mintues. So here instead we just write `4 hours` or `3 hours 40 mins`

CC: @nithindavid @pranavrajs

nithindavid commented:
@iamsivin Are we planning to address the comment by @scmmishra? Else I'll test this

iamsivin commented:
> @iamsivin Are we planning to address the comment by @scmmishra? Else I'll test this.

@nithindavid If we plan to do it that way, Shivam suggested. Then I can do it.
CC: @pranavrajs

pranavrajs commented:
@iamsivin Please go ahead and do it.

iamsivin commented:
@nithindavid @scmmishra Can you take a look at this PR? 

Here if the hour is greater than 3 now it will show `eg: We will be back online tomorrow at 11:00 PM` and if there is a day difference then it will show `eg: We will be back online on "dayName" at 11:00 PM`. 

If the hour is not greater than 3 then it will show `eg: We will be back online in 2 hours and 29 minutes`

muhsin-k commented:
I have a small suggestion here if the next available time is after today, show it tomorrow or the day name.
_We will be back online tomorrow at 12:00 AM_ should be  _We will be back online tomorrow_
_We will be back online on Monday at 12:00 AM_ should be  _We will be back online on Monday_ 

![CleanShot 2023-04-27 at 15 50 46](https://user-images.githubusercontent.com/12408980/234834742-eeb0ea83-dadd-4d31-8253-99ab0e3c679e.png)

iamsivin commented:
> We will be back online on Monday

@muhsin-k Done

iamsivin commented:
@pranavrajs @nithindavid @muhsin-k @scmmishra 
I have updated the PR in specific formats (added eg: as points below) and added support for the time zone.

1. If the next available time is below 3min - then `We will be back online in 5 minutes`
2. If the next available time is 2hrs and 33min - then `We will be back online in 2 hours and 35 minutes`
3. If the next available time is greater than 3hrs and within a day (eg: 5pm) - then `We will be back online at 05:00 PM`
4. If the next available time is the next day - then `We will be back online tomorrow`
5. If the next available time is 2 days after the current day (Tuesday) - then `We will be back online on Thursday`

nithindavid reviewed (changes requested):
Tested on Local and works as expected. Please see the minor code change suggested

iamsivin commented on app/javascript/widget/mixins/nextAvailabilityTime.js:
```diff
@@ -0,0 +1,256 @@
+export default {
+  data() {
+    return {
+      dayNames: {
```
@nithindavid Done https://github.com/chatwoot/chatwoot/pull/6902/commits/4d5a1e4733261dfb2ab369de8d3892e4652001a7

muhsin-k commented on app/javascript/widget/mixins/nextAvailabilityTime.js:
```diff
@@ -0,0 +1,256 @@
+    channelConfig() {
+      return window.chatwootWebChannel;
+    },
+    workingHoursEnabled() {
```
@iamsivin Do we need this computed property check everywhere since we are checking `workingHoursEnabled` before calling `timeLeftToBackInOnline` mixin?

![CleanShot 2023-05-02 at 20 35 37](https://user-images.githubusercontent.com/12408980/235706917-b9b0c8e1-b237-464b-b757-e1686b3aa755.png)

muhsin-k commented on app/javascript/widget/mixins/nextAvailabilityTime.js:
```diff
@@ -0,0 +1,256 @@
+  methods: {
+    getNextDay(day) {
+      // This code calculates the next day of the week based on the current day. If the current day is Saturday (6), then the next day will be Sunday (0).
+      return day === 6 ? 0 : day + 1;
```
```suggestion
      return (day + 1) % 7;
```

Please add the specs as well.

scmmishra commented on app/javascript/widget/mixins/nextAvailabilityTime.js:
```diff
@@ -0,0 +1,256 @@
+  timeSlotParse,
+  defaultTimeSlot,
+} from 'dashboard/routes/dashboard/settings/inbox/helpers/businessHour.js';
+const { utcToZonedTime } = require('date-fns-tz');
```
```suggestion
import { utcToZonedTime } from 'date-fns-tz';
```

ES6 imports style should be used

scmmishra commented on app/javascript/widget/mixins/nextAvailabilityTime.js:
```diff
@@ -0,0 +1,256 @@
+      }`;
+    },
+    hoursLeftValue() {
+      const {
+        hours: hoursLeft,
+        minutes: minutesLeft,
+      } = this.hoursAndMinutesBackInOnline;
```
Since we control the implementation of `hoursAndMinutesBackInOnline` we can send the values with `{ hoursLeft: 0, minutesLeft: 45}` directly instead of `hours` and `minutes` as keys. We will avoid the destructuring assignment

scmmishra commented on app/javascript/widget/mixins/nextAvailabilityTime.js:
```diff
@@ -0,0 +1,256 @@
+        this.timeSlot.day !== this.currentDay ? '' : `at ${this.timeSlot.from}`
+      }`;
+    },
+    hoursLeftValue() {
+      const {
+        hours: hoursLeft,
+        minutes: minutesLeft,
+      } = this.hoursAndMinutesBackInOnline;
+      const hourString = hoursLeft === 1 ? 'hour' : 'hours';
+      const roundedUpMinLeft = Math.ceil(minutesLeft / 5) * 5; // round up minutes to nearest multiple of 5
+      return `in ${hoursLeft} ${hourString}${
+        minutesLeft > 0 ? ` and ${roundedUpMinLeft} minutes` : ''
+      }`;
```
These methods can be collapsed into one. Assuming that the `hours` value from `this. hoursAndMinutesBackInOnline()` is 0

```js
    hoursAndMinutesLeft() {
      const { hours, minutes } = this.hoursAndMinutesBackInOnline;

      const timeLeftChars = ['in'];

      if (hours > 0) {
        const hourString = hours === 1 ? 'hour' : 'hours';
        timeLeftChars.push(`${hours} ${hourString}`);
      }

      if (minutes > 0) {
        const roundedUpMinLeft = Math.ceil(minutes / 5) * 5;
        timeLeftChars.push(`${roundedUpMinLeft} minutes`);
      }

      return timeLeftChars.join(' ');
    },
```

And the `hoursAndMinutesToBack` can look like this

```js
    hoursAndMinutesToBack() {
      const {
        hours: hoursLeft,
        minutes: minutesLeft,
      } = this.hoursAndMinutesBackInOnline;
      if (hoursLeft > 3) {
        return this.exactTimeInAmPm;
      }
      if (hoursLeft > 0 || minutesLeft > 0) {
        return this.hoursAndMinutesLeft;
      }
      return 'in some time';
    },
```

scmmishra commented on app/javascript/widget/mixins/nextAvailabilityTime.js:
```diff
@@ -0,0 +1,256 @@
+      );
+      return Number(currentDay);
+    },
+    currentDayOpenHour() {
+      return this.workingHoursEnabled
+        ? this.currentDayWorkingHours.open_hour
+        : null;
+    },
+    currentDayOpenMinute() {
+      return this.workingHoursEnabled
+        ? this.currentDayWorkingHours.open_minutes
+        : null;
+    },
+    currentDayCloseHour() {
+      return this.workingHoursEnabled
+        ? this.currentDayWorkingHours.close_hour
+        : null;
+    },
```
This can be collapsed into one, see suggestions below. 

```suggestion
    currentDatTimings() {
      const {
        open_hour: openHour,
        open_minutes: openMinute,
        close_hour: closeHour,
      } = this.currentDayWorkingHours ?? {};
      return {
        openHour,
        openMinute,
        closeHour,
      };
    },
```

However, a better way would be to use the `currentDayWorkingHours` computed directly, and have it return sane defaults. See second code block for that change, and then we can remove the three methods alltogether

```
 currentDayWorkingHours() {
      if (!this.workingHoursEnabled) {
        return {
          openHour: null,
          openMinute: null,
          closeHour: null,
        };
      }
      const {
        open_hour: openHour,
        open_minutes: openMinute,
        close_hour: closeHour,
      } = this.workingHours.find(slot => slot.day_of_week === this.currentDay);

      return {
        openHour,
        openMinute,
        closeHour,
      };
    },
```

scmmishra commented on app/javascript/widget/mixins/nextAvailabilityTime.js:
```diff
@@ -0,0 +1,256 @@
+        ? this.currentDayWorkingHours.close_hour
+        : null;
+    },
+    nextDayOpenHour() {
+      if (this.workingHoursEnabled) {
+        return this.nextDayWorkingHours.open_all_day
+          ? 0
+          : this.nextDayWorkingHours.open_hour;
+      }
+      return null;
+    },
+    nextDayOpenMinute() {
+      return this.workingHoursEnabled
+        ? this.nextDayWorkingHours.open_minutes
+        : null;
+    },
```
This can be collapsed into one as well

scmmishra commented on app/javascript/widget/mixins/nextAvailabilityTime.js:
```diff
@@ -0,0 +1,256 @@
+      }
+      return null;
+    },
+    newDateWithTimeZone() {
+      const date = new Date();
+      const timeZone = this.timeZoneValue;
+      const zonedDate = utcToZonedTime(date, timeZone);
+      return zonedDate;
+    },
+    presentHour() {
+      return this.newDateWithTimeZone.getHours();
+    },
+    presentMinute() {
+      return this.newDateWithTimeZone.getMinutes();
+    },
+    currentDay() {
+      const date = this.newDateWithTimeZone;
+      const day = date.getDay();
+      const currentDay = Object.keys(this.dayNames).find(
+        key => this.dayNames[key] === this.dayNames[day]
+      );
+      return Number(currentDay);
+    },
```
This can be moved out of the `computed` block altogether and can be added to the top-level. They don't need reactivity, neither do they use any values from the data props, except for `dayNames` which can be removed out too

scmmishra commented on app/javascript/widget/mixins/nextAvailabilityTime.js:
The check `if (this.workingHoursEnabled)` seems unnecessary in some places. Overall, we can reduce quite a few abstractions that don't add a lot of value

iamsivin commented:
@nithindavid @muhsin-k @scmmishra I have added fixes for all the comments. Can you take a look at this PR?

muhsin-k reviewed (approved):
All cases are working as expected! 
https://www.loom.com/share/d6f419e4fccf4a4480095fa3535c4786

pranavrajs commented on app/javascript/widget/components/ChatHeader.vue:
```diff
@@ -94,6 +95,12 @@ export default {
+      if (workingHoursEnabled) {
+        return this.isOnline
+          ? this.replyTimeStatus
+          : `We will be back online ${this.timeLeftToBackInOnline}`;
```
Please use i18n here.

iamsivin replied:
Done https://github.com/chatwoot/chatwoot/pull/6902/commits/2c0e179d2a22769ac7797a325de475ab2b416ce7

pranavrajs commented on app/javascript/widget/components/TeamAvailability.vue:
```diff
@@ -82,7 +83,9 @@ export default {
-        return this.outOfOfficeMessage;
+        return this.isOnline
+          ? this.replyTimeStatus
+          : `We will be back online ${this.timeLeftToBackInOnline}`;
```
Same as above.

pranavrajs replied:
Can't this be in the mixing as it is repeated in 2 different components?

iamsivin replied:
Done https://github.com/chatwoot/chatwoot/pull/6902/commits/2c0e179d2a22769ac7797a325de475ab2b416ce7

iamsivin commented:
@pranavrajs I have made the changes. Can you take a look?

scmmishra commented:
> @pranavrajs @nithindavid @muhsin-k @scmmishra I have updated the PR in specific formats (added eg: as points below) and added support for the time zone.
> 
> 1. If the next available time is below 3min - then `We will be back online in 5 minutes`
> 2. If the next available time is 2hrs and 33min - then `We will be back online in 2 hours and 35 minutes`
> 3. If the next available time is greater than 3hrs and within a day (eg: 5pm) - then `We will be back online at 05:00 PM`
> 4. If the next available time is the next day - then `We will be back online tomorrow`
> 5. If the next available time is 2 days after the current day (Tuesday) - then `We will be back online on Thursday`

Let's write test cases for all of these

scmmishra commented on app/javascript/widget/mixins/nextAvailabilityTime.js:
```diff
@@ -0,0 +1,234 @@
+    hoursAndMinutesLeft() {
+      const { hoursLeft, minutesLeft } = this.hoursAndMinutesBackInOnline;
+
+      const timeLeftChars = ['in'];
```
This won't work when translated

iamsivin replied:
Done https://github.com/chatwoot/chatwoot/pull/6902/commits/6e3648e99934120d6af00d97e7930edf6118d984

iamsivin commented:
A small update here

> @pranavrajs @nithindavid @muhsin-k @scmmishra I have updated the PR in specific formats (added eg: as points below) and added support for the time zone.
>
> 1. If the next available time is below 3min - then `We will be back online in 5 minutes`
> 2. If the next available time is 2hrs and 33min - then `We will be back online in 2 hours and 35 minutes`

Now it will return like this `We will be back online in 3 hours` (eg:2hrs and 33min then round up it to 3hrs)

> 3. If the next available time is greater than 3hrs and within a day (eg: 5pm) - then `We will be back online at 05:00 PM`
> 4. If the next available time is the next day - then `We will be back online tomorrow`
> 5. If the next available time is 2 days after the current day (Tuesday) - then `We will be back online on Thursday`

scmmishra commented on app/javascript/widget/mixins/nextAvailabilityTime.js:
```diff
@@ -0,0 +1,251 @@
+  data() {
+    return {
+      dayNames: [
+        'Sunday',
+        'Monday',
+        'Tuesday',
+        'Wednesday',
+        'Thursday',
+        'Friday',
+        'Saturday',
```
These need to be translated too

scmmishra commented on app/javascript/widget/mixins/nextAvailabilityTime.js:
```diff
@@ -0,0 +1,251 @@
+      }
+      return null;
+    },
+    generateRelativeTime(value, unit, languageCode) {
```
This can be moved out as a utility function elsewhere, same for `getnextDay`, `newDateWithTimezone`, basically any method that does not access the `data` properties from this mixin directly

scmmishra replied:
Let's just move the `generateRelativeTime` method outside, in `shared/helpers/DateHelper.js`

scmmishra commented on app/javascript/widget/mixins/nextAvailabilityTime.js:
```diff
@@ -0,0 +1,243 @@
+  },
+  methods: {
+    newDateWithTimeZone() {
+      const date = new Date();
+      const timeZone = this.timeZoneValue;
+      const zonedDate = utcToZonedTime(date, timeZone);
+      return zonedDate;
```
```suggestion
      return utcToZonedTime(new Date(), this.timeZoneValue);
```

scmmishra commented on app/javascript/widget/mixins/nextAvailabilityTime.js:
```diff
@@ -0,0 +1,243 @@
+      const zonedDate = utcToZonedTime(date, timeZone);
+      return zonedDate;
+    },
+    presentHour() {
+      return this.newDateWithTimeZone().getHours();
+    },
+    presentMinute() {
+      return this.newDateWithTimeZone().getMinutes();
+    },
+    currentDay() {
+      const date = this.newDateWithTimeZone();
+      const day = date.getDay();
+      const currentDay = Object.keys(this.dayNames).find(
+        key => this.dayNames[key] === this.dayNames[day]
+      );
+      return Number(currentDay);
+    },
```
These can be moved to computed as well

scmmishra commented on app/javascript/widget/mixins/nextAvailabilityTime.js:
```diff
@@ -0,0 +1,243 @@
+export default {
+  data() {
+    return {
+      dayNames: this.$t('DAY_NAMES'),
```
This is failing in tests

scmmishra commented on app/javascript/widget/mixins/nextAvailabilityTime.js:
```diff
@@ -0,0 +1,243 @@
+        this.dayDiff > 1 &&
+        this.presentHour() >= this.currentDayTimings.closeHour
+      ) {
+        return `on ${this.dayNameOfNextWorkingDay}`;
```
This is not translated

scmmishra commented on app/javascript/widget/mixins/nextAvailabilityTime.js:
```diff
@@ -0,0 +1,243 @@
+        return `on ${this.dayNameOfNextWorkingDay}`;
+      }
+      if (this.hoursAndMinutesBackInOnline.hoursLeft >= 24) {
+        return `tomorrow`;
```
This is not translated, we can use the relative time method

scmmishra reviewed (changes requested):
A few minor suggestions and this is good to merge.
